FEAT: Add Shared Schema recursion

WolfGreyDev opened the pull request:
This pull request introduces shared schema recursion to Syncify, allowing shared schemas to be nested within one another. This change adds more powerful block and setting composition, accompanied by necessary updates to the transform and caching logic, as well as minor fixes.

### Key Changes

1.  **Shared Schema Recursion (Inception)**
    * Shared schemas can now be used within other shared schemas for both `blocks` and `settings`. When a nested shared schema is detected, the transform process now performs additional checks to resolve the entire nested structure.

2.  **Expanded Schema Cache**
    * The schema cache has been updated to correctly traverse all nested shared schemas, collecting every file related to the referenced schemas. This now properly includes theme blocks, which were previously excluded.

3.  **Corrected Transform on Save**
    * **Previous Behavior**: When saving a shared schema, transforms were not applied to theme files. Instead, the files were directly uploaded, bypassing the output directory and causing a sync issue between local and remote theme files.
    * **New Behavior**: All related files are now properly transformed when a shared schema is saved *before* the upsert occurs, ensuring local and remote files remain synchronized. (See note 3 in "Additional Information").

4.  **Basic Loop Validation**
    * A check has been added to detect and warn when a shared schema directly loops on itself (e.g., `schema-a.a` references `schema-a.a`). The schema will be skipped from the transform to prevent an infinite loop.
    * **Limitation**: This does not prevent indirect loops (e.g., `schema-a.a` references `schema-b.a`, which references `schema-a.a`). Solving this would require a more significant design change.

5.  **Code Refactoring**
    * The code has been clarified by refactoring some nested logic into recursive functions. The internal settings logic was also improved for clearer functionality.

### How to Test

You can test this by creating complex nested structures.

**Settings:**
* Create multiple singleton schemas.
* Reference one or more of these singletons from within a "spread" schema.
* Reference one or more "spread" schemas from within a "collection" schema.
* This can be done for sections, theme blocks, or section blocks.

**Blocks:**
* Reference one or more singleton block schemas from within a collection schema.

**Both:**
* Reference a settings shared schema from within a block shared schema.

I have successfully tested nests of 7 layers deep with the following structure:
`Singleton -> Spread -> Spread -> Collection -> Block -> Block Collection -> Section`

### Additional Information

1.  **Builds vs. Shopify Validation**: It is now possible to add `settings` directly to `blocks` in your local files. Syncify will successfully build this, but the subsequent upload to Shopify will fail with a validation error. Please be aware that a successful local build does not guarantee a valid theme structure.

2.  **Recursive Logic**: The new logic is recursive and will always traverse down through nested schemas until it reaches a "singleton" (a schema with no further nests). These singletons are returned, and all settings are treated as spreads. Collections are effectively spreads nested within a settings property.

3.  **A Note on `watch` Mode Transforms**: The previous behavior of directly upserting shared schemas on save during `sy watch` (bypassing the output directory) was likely a design choice to speed up bulk uploads. I believe this created a poor developer experience as local and server files would become out of sync. To re-enable bulk uploading in the future, a flag could be added to the transform function to optionally skip the upsert, allowing it to be handled manually after all transforms are complete. This was not implemented as it is a larger design decision.

WolfGreyDev committed:
Add shared schema complilation to shared schemas

WolfGreyDev committed:
Add shared schema infinate looping conditional

WolfGreyDev committed:
Add recursive shared schema setting linked files to cache

WolfGreyDev committed:
Add shared schema to theme blocks and improved block handling

WolfGreyDev committed:
Refactored InjectSettings for clarity and performance improvement

WolfGreyDev committed:
Fix formatting

WolfGreyDev committed:
Add property overrides for shared schemas

WolfGreyDev commented:
### UPDATE 1

I have pushed an update to this PR to add property value overrides for settings and blocks.

### Key Changes

- Now we're able to control the values of properties from within a shared schema. This will allow us to pass a value all the way to the leaf node/schema item.
- Added a wildcard entry to daisy chain the values.
- Refactored `InjectBlocks` logic to match `InjectSettings`.
- Added new utility function `replaceAllOccurrences` to replace all occurrences of a sub-string inside another string.

### Examples

**Our base singleton file**
```jsonc
// colour-singleton.schema
{
  "setting": {
    "$description": "A colour dropdown",
    "type": "select",
    "id": "colour",
    "label": "Colour",
    "info": "Use the dropdown to select a colour",
    "default": "red",
    "options": [
      {
        "label": "Red",
        "value": "red"
      },
      {
        "label": "Blue",
        "value": "blue"
      },
      {
        "label": "Yellow",
        "value": "yellow"
      }
    ]
  }
}
```

**Overwriting the whole value**
<table>
  <thead>
    <tr>
      <th width="500px">Section Schema</th>
      <th width="500px">Shared Schema Result</th>
    </tr>
  </thead>
  <tbody>
  <tr width="600px">
      <td width="300px">

```jsonc
// section.liquid
{
  "$ref": "colour-singleton.setting",
  "id": "test"
}
```

```jsonc
// section.liquid
{
  "$ref": "colour-singleton.setting",
  "_settings": {
    "id": "test"
  }
}
```

</td>
<td width="300px">

```jsonc
// colour-singleton.schema
{
  "setting": {
    "type": "select",
    "id": "test",
    "label": "Colour",
    "info": "Use the dropdown to select a colour",
    "default": "red",
    "options": [ ]
  }
}
```

</td>
</tr>

  </tbody>
</table>

**Using a wildcard to prepend to the value**
<table>
  <thead>
    <tr>
      <th width="500px">Section Schema</th>
      <th width="500px">Shared Schema Result</th>
    </tr>
  </thead>
  <tbody>
  <tr width="600px">
      <td width="300px">

```jsonc
// section.liquid
{
  "$ref": "colour-singleton.setting",
  "id": "test_*"
}
```

```jsonc
// section.liquid
{
  "$ref": "colour-singleton.setting",
  "_settings": {
    "id": "test_*"
  }
}
```

</td>
<td width="300px">

```jsonc
// colour-singleton.schema
{
  "setting": {
    "type": "select",
    "id": "test_colour",
    "label": "Colour",
    "info": "Use the dropdown to select a colour",
    "default": "red",
    "options": [ ]
  }
}
```

</td>
</tr>

  </tbody>
</table>

**Removing a value**
<table>
  <thead>
    <tr>
      <th width="500px">Section Schema</th>
      <th width="500px">Shared Schema Result</th>
    </tr>
  </thead>
  <tbody>
  <tr width="500px">
      <td >

```jsonc
// section.liquid
{
  "$ref": "colour-singleton.setting",
  "info": null
}
```

```jsonc
// section.liquid
{
  "$ref": "colour-singleton.setting",
  "_settings": {
    "info": null
  }
}
```

</td>
<td>

```jsonc
// colour-singleton.schema
{
  "setting": {
    "type": "select",
    "id": "colour",
    "label": "Colour",
    "default": "red",
    "options": [ ]
  }
}
```

</td>
</tr>

  </tbody>
</table>

**Using this is spreads**
<table>
  <thead>
    <tr>
      <th width="500px">Section Schema</th>
      <th width="500px">Shared Schema Result</th>
    </tr>
  </thead>
  <tbody>
  <tr width="500px">
      <td >

```jsonc
// section.liquid
{
  "$ref": "colours.spread",
  "_settings": {
    "id": "section_*",
    "default": "blue"
  }
}
```

```jsonc
// colours.schema
{
  "spread": [
    {
      "$ref": "colour-singleton.setting",
      "_settings": {
        "id": "background_*",
        "label": "Background *",
        "info": null
      }
    },
    {
      "$ref": "colour-singleton.setting",
      "_settings": {
        "id": "text_*",
        "label": "* (Text)",
        "default": "yellow"
      }
    }
  ]
}
```

</td>
<td>

```jsonc
// colour-singleton.schema
{
  "setting": {
    "type": "select",
    "id": "section_background_colour",
    "label": "Background Colour",
    "default": "blue",
    "options": [ ]
  }
}
// colour-singleton.schema
{
  "setting": {
    "type": "select",
    "id": "section_text_colour",
    "label": "Colour (Text)",
    "info": "Use the dropdown to select a colour",
    "default": "yellow",
    "options": [ ]
  }
}
```

</td>
</tr>

  </tbody>
</table>

**Working with blocks and block settings**
<table>
  <thead>
    <tr>
      <th width="500px">Section Schema</th>
      <th width="500px">Shared Schema Result</th>
    </tr>
  </thead>
  <tbody>
  <tr width="500px">
      <td >

```jsonc
// section.liquid
"blocks": [
  {
    "$ref": "colour-block-singleton.setting",
    "_block": {
      "name": "block_*"
    },
    "_settings": {
      "id": "block_settings_*"
    }
  }
]
```

</td>
<td>

```jsonc
// colour-block-singleton.schema
{
  "setting": {
    "$description": "Info about block",
    "name": "block_foo",
    "type": "foo",
    "settings": [
      {
        "type": "text",
        "id": "block_settings_title",
        "label": "Example",
        "default": "Hello World"
      },
      {
        "type": "textarea",
        "id": "block_settings_body",
        "label": "Body"
      }
    ]
  }
}
```

</td>
</tr>

  </tbody>
</table>

### Additional Information

1. For now, it's strictly limited to string type values. I'm unsure if there would be a need to dynamically populate other properties as they're usually closely tied with the singleton's expected result. Values include:
    - **Settings**: 'id', 'label', 'info', 'visible_if', 'default'. 
    Note, `visible_if` doesn't allow for a wildcard entry as it contains liquid and therefore requires the wildcard value `*` to be available for math evaluations. This setting is best used in spreads which call a singleton. Beyond that, I don't think it's feasible to daisy chain values.
    - **Blocks**: 'name', 'type'

2. Adding `null` to any of these properties at any stage of the recursion, will remove it from the final output.

3. As per the examples, we're able to add the properties directly to the schema item. However, for consistency, `_settings` and `_blocks` serve the same function and should be used for advanced schema nesting. It's my recommendation that `_settings` and `_blocks` be the recommended and default option when using this functionality. Note, a block that requires settings property value changes MUST use the `_settings` property.

WolfGreyDev committed:
Add options property overrides for select and radio types in shared schemas

WolfGreyDev committed:
Fix shared schema cache for recursive branch schemas

WolfGreyDev commented:
### UPDATE 2

I have pushed an update to this PR to add property value overrides for options, specifically targeting select and radio types.

### Key Changes

- Adding an `options` parameter to a recursive shared schema will pass the values to the leaf node.
- The options can have either `label` and/or `value`.
- FIXED: branch schemas not being added to cache. This caused an issue that when a branch schema changed, it did not update the original referrer schema/section/block.

### Examples

**Our base singleton file**
```jsonc
// colour-singleton.schema
{
  "setting": {
    "$description": "A colour dropdown",
    "type": "select",
    "id": "colour",
    "label": "Colour",
    "info": "Use the dropdown to select a colour",
    "default": "red",
    "options": [
      {
        "label": "Red",
        "value": "red"
      },
      {
        "label": "Blue",
        "value": "blue"
      },
      {
        "label": "Yellow",
        "value": "yellow"
      }
    ]
  }
}
```

**Overwriting the whole value**
<table>
  <thead>
    <tr>
      <th width="500px">Section Schema</th>
      <th width="500px">Shared Schema Result</th>
    </tr>
  </thead>
  <tbody>
  <tr width="600px">
      <td width="300px">

```jsonc
// section.liquid
{
  "$ref": "colour-singleton.setting",
  "options": {
    "label": "foo"
  }
}
```

```jsonc
// section.liquid
{
  "$ref": "colour-singleton.setting",
  "_settings": {
    "options": {
      "label": "foo"
    }
  }
}
```

</td>
<td width="300px">

```jsonc
// colour-singleton.schema
{
  "setting": {
    "type": "select",
    "id": "test",
    "label": "Colour",
    "info": "Use the dropdown to select a colour",
    "default": "red",
    "options": [
      {
        "label": "foo",
        "value": "red"
      },
      {
        "label": "foo",
        "value": "blue"
      },
      {
        "label": "foo",
        "value": "yellow"
      }
    ]
  }
}
```

</td>
</tr>

  </tbody>
</table>

**Using a wildcard to prepend to the value**
<table>
  <thead>
    <tr>
      <th width="500px">Section Schema</th>
      <th width="500px">Shared Schema Result</th>
    </tr>
  </thead>
  <tbody>
  <tr width="600px">
      <td width="300px">

```jsonc
// section.liquid
{
  "$ref": "colour-singleton.setting",
  "options": {
    "label": "foo *"
  }
}
```

```jsonc
// section.liquid
{
  "$ref": "colour-singleton.setting",
  "_settings": {
    "options": {
      "label": "foo *"
    }
  }
}
```

</td>
<td width="300px">

```jsonc
// colour-singleton.schema
{
  "setting": {
    "type": "select",
    "id": "colour",
    "label": "Colour",
    "info": "Use the dropdown to select a colour",
    "default": "red",
    "options": [
      {
        "label": "foo Red",
        "value": "red"
      },
      {
        "label": "foo Blue",
        "value": "blue"
      },
      {
        "label": "foo Yellow",
        "value": "yellow"
      }
    ]
  }
}
```

</td>
</tr>

  </tbody>
</table>

**Adding to values**
<table>
  <thead>
    <tr>
      <th width="500px">Section Schema</th>
      <th width="500px">Shared Schema Result</th>
    </tr>
  </thead>
  <tbody>
  <tr width="500px">
      <td >

```jsonc
// section.liquid
{
  "$ref": "colour-singleton.setting",
  "default": "bar *",
  "options": {
    "label": "foo *",
    "value": "bar *"
  }
}
```

```jsonc
// section.liquid
{
  "$ref": "colour-singleton.setting",
  "_settings": {
    "default": "bar *",
    "options": {
      "label": "foo *",
      "value": "bar *"
    }
  }
}
```

</td>
<td>

```jsonc
// colour-singleton.schema
{
  "setting": {
    "type": "select",
    "id": "colour",
    "label": "Colour",
    "info": "Use the dropdown to select a colour",
    "default": "bar red",
    "options": [
      {
        "label": "foo Red",
        "value": "bar red"
      },
      {
        "label": "foo Blue",
        "value": "bar blue"
      },
      {
        "label": "foo Yellow",
        "value": "bar yellow"
      }
    ]
  }
}
```

</td>
</tr>

  </tbody>
</table>

### Additional Information

1. Syncify will not handle any validation within the shared schema logic. Instead, we'll continue to rely on the upsert validation to warn the user if there are any validation errors. This means, so long as the schema is valid and can build, Syncify will build the schema as is.

2. When adding a `value` mutation to the `options`, you _should_ set the `default` property to match your `value`. If no `default` is provided, Syncify will process this but it will fail when upserting due to a validation error. This could be a potential feature update but I want to stay away from magically modifying any values behind the scenes.

3. The group property has been specifically left out of being mutated. I don't see any situation where you would want to build a group label however, if the use case becomes more obvious, this could be added at a later date.

WolfGreyDev committed:
Fix unused imports

WolfGreyDev committed:
Add shared schema and related files to cache in watch mode

WolfGreyDev committed:
Fix theme upsert error when deleting shared schema

WolfGreyDev committed:
Add shared schema blocks to cache in watch mode

WolfGreyDev commented:
### UPDATE 3

I have pushed an update to this PR to fix old issues related to shared schemas and watch mode.

### The Issue/s

1. When adding a new shared schema in watch mode, the initial file creation is handled by Syncify and skipped since the file is empty. However, if the newly created shared schema file is saved with settings/blocks, then Syncify would crash as it tries to convert an `undefined` to an array.

2. Sections or Blocks that reference a shared schema after the initial cache build are never included in a backwards save from the shared schema. This is due to the shared schema cache not having the section/block item in it's set.

3. To "fix" these issues, we would normally need to leave the watch mode and re-build the theme to update the cache.

4. When deleting a shared schema, Syncify will try to remove the file from the online theme. This file never exists and therefore throws an error.

### Key Changes

1. Dynamic cache updates while in watch mode

    - When a shared schema is added in watch mode, if it does not exist in the cache , it will create an empty set in the cache. This stops the issue of creating an array on the `undefined` property.
    
    - When a section or block is saved, if the shared schema exists and the file isn't in the set, it will add the file for future saves.
    
2. Deletes are handled gracefully

    - When deleting a shared schema, we now check if the file is of a schema type. If it is not a schema, the operation continues as normal. If the file is a schema, we warn the user of the files (sections/blocks) that have been affected and remove the schema from the cache.

### Additional Notes

- The dynamic cache will only ever add files (sections/blocks) to the schema cache. This means that we pick up any new additions while continuing to work as normal. The side effect of not dynamically removing these files is when a shared schema is saved, if at any point during this watch session, then it will be included in the save and upsert.

    This is a design decision for performance. If we were to remove files, we'd need to query sections/blocks/schema files on every action to constantly rebuild the cache which defeats its purpose. Since a shared schema save triggers a liquid transform, the transform will not know that it previously had a shared schema reference and thus cannot remove itself.
    
    If the cache becomes massively outdated, a rebuild will flush those orphaned entries. However, the user can continue to use watch mode with these orphaned entries as the liquid transform will still successfully process the sections/blocks.
    
- When deleting a shared schema, Syncify always processed each file individually. I was unable to trigger a bulk file action. This may mean that in the future, if the bulk file action is ever called, we'll need to add the same check as the single file change.

panoply commented:
Reviewing this tonight, amazing work here from what I have seen so far. I am not sure if caching registers correctly but I haven't looked close enough, so grain of salt me. We will need to tie this in with the JSON IntelliSense logic within vscode-liquid

panoply commented:
Can you share a `tests/cases/shared-schema` theme for me here with these changes so I can play?

WolfGreyDev committed:
Add flushing of files from schema cache before rebuilding

WolfGreyDev committed:
Add e2e test for shared schema nesting

WolfGreyDev committed:
Add placeholder has allowedProp for schema property override

WolfGreyDev committed:
Fix null values being ignored for schema property overrides

WolfGreyDev committed:
Fix radio option values being ignored for schema property overrides

WolfGreyDev committed:
Add automatic resolving of default property for select and radio settings for schema property overrides

WolfGreyDev committed:
Add e2e test for property overrides in shared schemas

WolfGreyDev commented:
### UPDATE 4

I have pushed an update to this pr to fix a few bugs, add e2e tests cases and add a convenience feature.

### Key Changes

1. **Smarter Shared Schema Cache**

    - This PR added the feature for Shared Schemas to add any file (section/block) to the cache for easy access in the future. This presented small issue where if the section/block file removed the Shared Schema reference, the cache wouldn't know about it. As such, if the Shared Schema was saved, it could iterate over files that no longer had a reference to that Shared Schema. It was not a breaking issue, just resulted in unnecessary calls to the Shopify store and processing.
    - Now when a section/block file is transformed, it will remove itself from the Shared Schema cache and add itself back to any Shared Schema references it has. This results in an up to date Shared Schema cache and if a Shared Schema is saved, only the sections/blocks that have a reference to this Shared Schema will be transformed and upserted.

2. **Allow Placeholder Property Overrides**

    - The `placeholder` property was unintentionally left out of the allowed props for property overrides. This will now  operate just like any other string property override.

3. **Auto Resolve Default Property For Select/Radio Overrides**

    - This PR allows for `label` and `value` overrides for `options`, specifically for `select` and `radio` type settings. When in isolation, this feature functioned as it should, requiring the user to add a `default` value that matched the final output in order to upsert a valid schema. In practice this resulted in additional settings inheriting this value and could cause issues and unwanted values.
    - Now when the `options` property contains a `value` property, at any point during the nesting process, the `default` value is also updated so the final result will match an option. The `default` property will be ignored for this type.
    - Additionally, if a `default` property is added to the schema, this will only effect other string based types. There is no support for integer.

4. **Bug Fixes**

    - `null` values weren't updated to account for the new method of output into a singleton. Now the `null` value will be added to the overrides and passed through to the final singleton where it'll be removed from the final output.
    - When adding `options` property for overriding, when used with a `radio` type, we'd remove the `options` overrides before continuing through the nested logic. Now we've added `radio` as a valid type for `options` and it won't prematurely exit.

5. **e2e Tests**

    - This PR now has 2 e2e tests under `tests/e2e/shared-schemas`
        - `nested-schemas` is specifically for Shared Schema nesting and
        - `property-overrides` deals with the property override functionality
    - 2 separate tests were chosen since it's easier to debug when adding more features to each.
    - **Note**: the `property-override` tests requires the `nested-schemas` to be passing. I have added notes to the test cases (sections) for expected results.

WolfGreyDev committed:
Fix issue with default property values not inheriting the correct value
